File: elixir/rest_api/lib/rest_api/store.ex
defmodule RestApi.Store do
  use Agent

  alias RestApi.Models.{User, Post}

  def start_link(_),
    do:
      Agent.start_link(fn -> %{users: %{}, posts: %{}, user_id: 0, post_id: 0} end,
        name: __MODULE__
      )

  def create_user(data) do
    Agent.get_and_update(__MODULE__, fn state ->
      id = state.user_id + 1
      now = DateTime.utc_now() |> DateTime.to_iso8601()

      user = %User{
        id: id,
        name: data["name"],
        email: data["email"],
        role: data["role"] || "user",
        created_at: now,
        updated_at: now
      }

      state = %{state | users: Map.put(state.users, id, user), user_id: id}
      {user, state}
    end)
  end

  def get_user(id), do: Agent.get(__MODULE__, &Map.get(&1.users, id))

  def list_users(page, per_page, search) do
    Agent.get(__MODULE__, fn state ->
      items = Map.values(state.users)

      items =
        if search do
          q = String.downcase(search)

          Enum.filter(
            items,
            &(String.contains?(String.downcase(&1.name), q) or
                String.contains?(String.downcase(&1.email), q))
          )
        else
          items
        end

      RestApi.Models.paginate(items, page, per_page)
    end)
  end

  def update_user(id, data) do
    Agent.get_and_update(__MODULE__, fn state ->
      case Map.get(state.users, id) do
        nil ->
          {nil, state}

        user ->
          now = DateTime.utc_now() |> DateTime.to_iso8601()
          updated = %{user | updated_at: now}
          updated = if data["name"], do: %{updated | name: data["name"]}, else: updated
          updated = if data["email"], do: %{updated | email: data["email"]}, else: updated
          updated = if data["role"], do: %{updated | role: data["role"]}, else: updated
          {updated, %{state | users: Map.put(state.users, id, updated)}}
      end
    end)
  end

  def delete_user(id) do
    Agent.get_and_update(__MODULE__, fn state ->
      if Map.has_key?(state.users, id) do
        posts = state.posts |> Enum.reject(fn {_, p} -> p.author_id == id end) |> Map.new()
        {true, %{state | users: Map.delete(state.users, id), posts: posts}}
      else
        {false, state}
      end
    end)
  end

  def count_user_posts(user_id) do
    Agent.get(__MODULE__, fn state ->
      Enum.count(state.posts, fn {_, p} -> p.author_id == user_id end)
    end)
  end

  def create_post(author_id, data) do
    Agent.get_and_update(__MODULE__, fn state ->
      id = state.post_id + 1
      now = DateTime.utc_now() |> DateTime.to_iso8601()

      post = %Post{
        id: id,
        author_id: author_id,
        title: data["title"],
        body: data["body"],
        published: data["published"] || false,
        created_at: now,
        updated_at: now
      }

      state = %{state | posts: Map.put(state.posts, id, post), post_id: id}
      {post, state}
    end)
  end

  def get_post(id), do: Agent.get(__MODULE__, &Map.get(&1.posts, id))

  def list_posts(page, per_page, author_id, published_only) do
    Agent.get(__MODULE__, fn state ->
      items = Map.values(state.posts)
      items = if author_id, do: Enum.filter(items, &(&1.author_id == author_id)), else: items
      items = if published_only, do: Enum.filter(items, & &1.published), else: items
      RestApi.Models.paginate(items, page, per_page)
    end)
  end

  def update_post(id, data) do
    Agent.get_and_update(__MODULE__, fn state ->
      case Map.get(state.posts, id) do
        nil ->
          {nil, state}

        post ->
          now = DateTime.utc_now() |> DateTime.to_iso8601()
          updated = %{post | updated_at: now}
          updated = if data["title"], do: %{updated | title: data["title"]}, else: updated
          updated = if data["body"], do: %{updated | body: data["body"]}, else: updated

          updated =
            if Map.has_key?(data, "published"),
              do: %{updated | published: data["published"]},
              else: updated

          {updated, %{state | posts: Map.put(state.posts, id, updated)}}
      end
    end)
  end

  def delete_post(id) do
    Agent.get_and_update(__MODULE__, fn state ->
      if Map.has_key?(state.posts, id) do
        {true, %{state | posts: Map.delete(state.posts, id)}}
      else
        {false, state}
      end
    end)
  end
end

File: python/rest_api/rest_api/store.py
from datetime import datetime

from .models import (
    PaginatedResponse,
    Post,
    PostCreate,
    PostUpdate,
    User,
    UserCreate,
    UserUpdate,
)


class Store:
    def __init__(self) -> None:
        self.users: dict[int, User] = {}
        self.posts: dict[int, Post] = {}
        self._user_id = 0
        self._post_id = 0

    def _next_user_id(self) -> int:
        self._user_id += 1
        return self._user_id

    def _next_post_id(self) -> int:
        self._post_id += 1
        return self._post_id

    def _post_count(self, user_id: int) -> int:
        return sum(1 for p in self.posts.values() if p.author_id == user_id)

    def _with_post_count(self, user: User) -> User:
        return user.model_copy(update={"post_count": self._post_count(user.id)})

    def create_user(self, data: UserCreate) -> User:
        now = datetime.now()
        user = User(id=self._next_user_id(), **data.model_dump(), created_at=now, updated_at=now)
        self.users[user.id] = user
        return self._with_post_count(user)

    def get_user(self, user_id: int) -> User | None:
        user = self.users.get(user_id)
        return self._with_post_count(user) if user else None

    def list_users(self, page: int, per_page: int, search: str | None) -> PaginatedResponse[User]:
        items = list(self.users.values())
        if search:
            query = search.lower()
            items = [u for u in items if query in u.name.lower() or query in u.email.lower()]
        total = len(items)
        start = (page - 1) * per_page
        items = [self._with_post_count(u) for u in items[start : start + per_page]]
        return PaginatedResponse(
            items=items,
            total=total,
            page=page,
            per_page=per_page,
            pages=(total + per_page - 1) // per_page,
        )

    def update_user(self, user_id: int, data: UserUpdate) -> User | None:
        user = self.users.get(user_id)
        if user is None:
            return None
        updates = data.model_dump(exclude_unset=True)
        updated = user.model_copy(update={**updates, "updated_at": datetime.now()})
        self.users[user_id] = updated
        return self._with_post_count(updated)

    def delete_user(self, user_id: int) -> bool:
        if self.users.pop(user_id, None) is None:
            return False
        self.posts = {k: v for k, v in self.posts.items() if v.author_id != user_id}
        return True

    def create_post(self, author_id: int, data: PostCreate) -> Post:
        now = datetime.now()
        post = Post(
            id=self._next_post_id(),
            author_id=author_id,
            **data.model_dump(),
            created_at=now,
            updated_at=now,
        )
        self.posts[post.id] = post
        return post

    def get_post(self, post_id: int) -> Post | None:
        return self.posts.get(post_id)

    def list_posts(
        self,
        page: int,
        per_page: int,
        author_id: int | None,
        published_only: bool,
    ) -> PaginatedResponse[Post]:
        items = list(self.posts.values())
        if author_id is not None:
            items = [p for p in items if p.author_id == author_id]
        if published_only:
            items = [p for p in items if p.published]
        total = len(items)
        start = (page - 1) * per_page
        items = items[start : start + per_page]
        return PaginatedResponse(
            items=items,
            total=total,
            page=page,
            per_page=per_page,
            pages=(total + per_page - 1) // per_page,
        )

    def update_post(self, post_id: int, data: PostUpdate) -> Post | None:
        post = self.posts.get(post_id)
        if post is None:
            return None
        updates = data.model_dump(exclude_unset=True)
        updated = post.model_copy(update={**updates, "updated_at": datetime.now()})
        self.posts[post_id] = updated
        return updated

    def delete_post(self, post_id: int) -> bool:
        return self.posts.pop(post_id, None) is not None


store = Store()
